tron/serialize/runstate/dynamodb_state_store.py

```diff
@@ -47,38 +48,42 @@ def restore(self, keys) -> dict:
         vals = self._merge_items(first_items, remaining_items)
         return vals
 
-    def _get_items(self, keys: list) -> object:
-        items = []
+    def chunk_keys(self, keys: list) -> list:
+        """Generates the cand keys list to be used to read from DynamoDB"""
+        # have a for loop here for all the key chunks we want to go over
+        cand_keys_chunks = []
         for i in range(0, len(keys), 100):
-            count = 0
-            cand_keys = keys[i : min(len(keys), i + 100)]
-            while True:
-                resp = self.client.batch_get_item(
-                    RequestItems={
-                        self.name: {
-                            "Keys": cand_keys,
-                            "ConsistentRead": True,
-                        },
-                    },
-                )
-                items.extend(resp["Responses"][self.name])
-                if resp["UnprocessedKeys"].get(self.name) and count < 10:
-                    cand_keys = resp["UnprocessedKeys"][self.name]["Keys"]
-                    count += 1
-                elif count >= 10:
-                    error = Exception(
-                        f"tron_dynamodb_restore_failure: failed to retrieve items with keys \n{cand_keys}\n from dynamodb\n{resp}"
+            # chunks of 100 keys will be in this list
+            cand_keys_chunks.append(keys[i : min(len(keys), i + 100)])
+        return cand_keys_chunks
+
+    def _get_items(self, table_keys: list) -> object:
+        items = []
+        # precompute the cand_keys and then all we gotta do is submit stuff to the thread pool using the precomputed keys
+        cand_keys_list = self.chunk_keys(table_keys)
+        count = 0
+        while count < len(cand_keys_list):
```

Comment:
i'm not quite sure we're doing the same thing as before with the new code: if i'm reading the old code correctly, we did 10 attempts per-chunk to get all keys for a given chunk

Reply:
yea, my code is giving one chance/attempt to get keys and unprocessed keys for a chunk. I'm not sure why we give 10 attempts? why not more and why not less?

Reply:
hmm rn though if we don't get the keys by first attempt we miss on them. I'm basically looping through the list of chunks size and attempting to get key's response on the first time

Reply:
10 was likely picked as an arbitrary # if i had to guess :)

my comment was more around this doing `len(cand_keys_list)` attempts now (if i'm reading this right), which would be a very small number in small clusters and potentially large number in big clusters